alf/bin/grid_search.py

```diff
@@ -77,16 +92,33 @@ def __init__(self, conf_file):
             param_keys = []
             param_values = []
             conf = json.loads(f.read())
+            assert 'parameters' in conf, "JSON must contain 'parameters' key!"
             for key, value in conf['parameters'].items():
                 if isinstance(value, str):
                     value = eval(value)
                 param_keys.append(key)
                 param_values.append(value)
 
-        self._desc = conf['desc']
+        # check conf keys
+        for k in conf.keys():
+            assert k in self._all_keys_, "Invalid conf key: %s" % k
+
+        self._desc = conf.get('desc', "Grid search")
         self._param_keys = param_keys
         self._param_values = param_values
-        self._max_worker_num = conf['max_worker_num']
+        self._max_worker_num = conf.get('max_worker_num', 1)
+        self._use_gpu = conf.get("use_gpu", False)
```

Comment:
should default to whether gpu is available.

Reply:
Checking whether gpu is available or not will bring in additional dependency (other lib) here, because if we check gpu with TF here in the beginning, then it will trigger a process so that we cannot set CUDA_VISIBLE_DEVICES later for a subprocess any more.